bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/PageImpl.java

```diff
@@ -0,0 +1,451 @@
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.wrappers.SlingHttpServletRequestWrapper;
+import org.apache.sling.models.annotations.Exporter;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.ScriptVariable;
+
+import com.adobe.cq.export.json.ContainerExporter;
+import com.adobe.cq.export.json.ExporterConstants;
+import com.adobe.cq.export.json.hierarchy.type.HierarchyTypes;
+import com.adobe.cq.wcm.core.components.internal.Utils;
+import com.adobe.cq.wcm.core.components.models.Page;
+import com.day.cq.wcm.api.Template;
+import com.day.cq.wcm.api.TemplatedResource;
+import com.day.cq.wcm.api.components.ComponentContext;
+import com.day.cq.wcm.api.designer.Style;
+import com.day.cq.wcm.api.policies.ContentPolicy;
+import com.day.cq.wcm.api.policies.ContentPolicyManager;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = {Page.class, ContainerExporter.class}, resourceType = PageImpl.RESOURCE_TYPE)
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME, extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class PageImpl extends com.adobe.cq.wcm.core.components.internal.models.v2.PageImpl implements Page {
+
+    protected static final String RESOURCE_TYPE = "core/wcm/components/page/v3/page";
+
+    /**
+     * Flags the child pages. Optionally available as a request attribute
+     */
+    private static final String ATTR_IS_CHILD_PAGE = "com.adobe.cq.wcm.core.components.internal.models.HierarchyPage.isChildPage";
+
+    /**
+     * Is the current model to be considered as a model root
+     */
+    private static final String PR_IS_ROOT = "isRoot";
+
+    /**
+     * URL extension specific to the Sling Model exporter
+     */
+    private static final String URL_MODEL_EXTENSION = ".model.json";
+
+    /**
+     * Request attribute key of the request page entry point
+     */
+    private static final String ATTR_HIERARCHY_ENTRY_POINT_PAGE = "com.adobe.cq.wcm.core.components.internal.models.HierarchyPage.entryPointPage";
+
+    @ScriptVariable
+    private Resource resource;
+
+    private Map<String, ? extends Page> childPages = null;
+
+    private com.day.cq.wcm.api.Page rootPage = null;
+
+    @Nullable
+    @Override
+    public String getExportedHierarchyType() {
+        return HierarchyTypes.PAGE;
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, ? extends Page> getExportedChildren() {
+        if (childPages == null) {
+            childPages = getChildPageModels();
+        }
+
+        return childPages;
+    }
+
+    @Nonnull
+    @Override
+    public String getExportedPath() {
+        return currentPage.getPath();
+    }
+
+    @Nullable
+    @Override
+    public String getRootUrl() {
+        if (currentStyle != null && currentStyle.containsKey(PR_IS_ROOT)) {
+            return getModelUrl(request, currentPage);
+        }
```

Comment:
This logic should be part of the logic in `getRootPage()` and should not be repeated here.